KIPs/kip-114.md

```diff
@@ -0,0 +1,324 @@
+
+An example of such use-cases is the block proposer selection. An exemplary requirement can be that the next block's proposer remains undisclosed until the latest block has been finalized, or that the proposer cannot tamper with deciding the next block's proposer.
+
+In this proposal, new header fields `randomReveal` and `mixHash` are introduced. `randomReveal` is the signature of the proposer, and `randomReveal`s are mixed to compute `mixHash`, which users can use for a part of the random source. `mixHash` is unpredictable, verifiable, and dependable:
+
+- (Unpredictability) the signature is random and so is `mixHash`.
+- (Verifiability) the signature can be verified with the public key of the proposer, and `mixHash` is computed in a deterministic manner.
+- (Dependability) `mixHash` only relies on the validators.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.
+
+The terms hash_to_point (hereinafter referred to as hash), public key, proof-of-possession (hereinafter referred to as pop) are from the ciphersuite [BLS_SIG_BLS12381G2_XMD:SHA-256_SSWU_RO_POP\_](https://www.ietf.org/archive/id/draft-irtf-cfrg-bls-signature-05.html#section-4.2.3).
+
+### Parameters
+
+| Constant     | Value |
+| ------------ | ----- |
+| `FORK_BLOCK` | TBD   |
+
+### Block Structure
+
+Beginning with `FORK_BLOCK`, client software must set the fields in a block header:
+
+- `randomReveal`: 96 bytes. It must be specified in the 15th field (0-indexed) of a block header.
+
+- `mixHash`: 32 bytes. It must be specified in the 16th field (0-indexed) of a block header.
+
+### Block Processing
+
+#### Block Proposal
+
+Beginning with `FORK_BLOCK`, the proposer must fill the new fields in the block proposal:
+
+- `randomReveal`: The proposer must sign the hash of the proposing block number.
+
+- `mixHash`: The proposer must XOR the keccak256 hash of `randomReveal` and the previous block's `mixHash`. If the previous block is before `FORK_BLOCK`, then the latter must be 32 bytes of zero.
+
+#### Validation
+
+Beginning with `FORK_BLOCK`, validators must validate the new fields:
+
+- `randomReveal`
+
+  - The validators must fetch the public key and the pop of the proposer by `getInfo(cnNodeId)` of the BLS registry standardized by [KIP-113](https://github.com/klaytn/kips/blob/main/KIPs/kip-113.md#validation).
+  - The validators must verify the validity of the proposer public key using pop as mentioned in [here](https://github.com/klaytn/kips/blob/main/KIPs/kip-113.md#validation).
+  - The validators must verify if `randomReveal` is signed by the proposer public key.
+
+- `mixHash`: The validators must verify the validity of the proposed `mixHash` by calculating `mixHash` using the aforementioned `mixHash` algorithm.
+
+### Pseudocode
+
+See the following pseudocode that implements the block processing in python:
```

Comment:
hmm actually this is not a pseudo code. It is a workable code with aux functionalities. Can we make it concise so that any readers can easily find the core algorithm?

Reply:
Added concise pseudocode, moved this to implementation section.